include/RevProc.h

```diff
@@ -190,8 +184,23 @@ class RevProc{
   ///  RevProc
   void ExternalReleaseHart(RevProcPasskey<RevCoProc>, uint16_t HartID);
   //------------- END External Interface -------------------------------
-  // Find a Hart to assign ThreadID to
-  void AssignThread(uint32_t ThreadID);
+
+  ///< RevProc: Used for loading a software thread into a RevHart
+  void AssignThread(std::unique_ptr<RevThread> ThreadToAssign);
+
+  ///< RevProc:
+  void UpdateStatusOfHarts();
+
+  ///< RevProc: Returns the id of an idle hart (or _INVALID_HART_ID_ if none are idle)
+  unsigned FindIdleHartID() const ;
+
+  ///< RevProc: Returns true if all harts are available (ie. There is nothing executing on this Proc)
+  /// @DAVE: Can you confirm that using this function to determine if a Proc can be disabled (called in RevCPU)
+  ///        will work with the CoProc?
+  size_t HasNoBusyHarts() const { return IdleHarts == ValidHarts; }
```

Comment:
I'll answer your question in a few ways... hopefully this helps.  

A thread should go to DONE state (as I understand it) once all instructions have been retired (similar to the `return` from `main`) and the `RevCoProc::IsDone() `returns true. 

A thread could certainly be Idle if it is waiting on a memory request or other dependency, this would be independent of the CoProc state.  It is also possible that the CoProc has forced a Hart to stall on Proc

A Proc could possibly be safely be disabled regardless of CoProc state **if** the CoProc is being clocked independently.  That is still a risky proposition as the CoProc can see and manipulate state within Proc and that is really not possible if the Proc is not being clocked.  The best strategy is to, as the general case, assume that the Proc cannot be disabled unless CoProc::IsDone() returns true.

Reply:
> Why `size_t` return type instead of `bool`?

Isn't that a great question 😂

Reply:
> I'll answer your question in a few ways... hopefully this helps.
> 
> A thread should go to DONE state (as I understand it) once all instructions have been retired (similar to the `return` from `main`) and the `RevCoProc::IsDone() `returns true.
> 
> A thread could certainly be Idle if it is waiting on a memory request or other dependency, this would be independent of the CoProc state. It is also possible that the CoProc has forced a Hart to stall on Proc
> 
> A Proc could possibly be safely be disabled regardless of CoProc state **if** the CoProc is being clocked independently. That is still a risky proposition as the CoProc can see and manipulate state within Proc and that is really not possible if the Proc is not being clocked. The best strategy is to, as the general case, assume that the Proc cannot be disabled unless CoProc::IsDone() returns true.

So a thread won't be removed from a Hart until all Scoreboard entries are false. When the CoProc flags a scoreboard, is it independent of the Hart's scoreboard? Also is there state beyond just the scoreboard that the CoProc can manipulate/see into?

Reply:
At the moment the scoreboard and the register file are the only things the CoProc can manipulate.  It is possible, however, for a Hart to have a clear scoreboard and the CoProc to have outstanding work

Reply:
Enough said... I'll adjust the check in RevCPU